rllib/core/learner/learner.py

```diff
@@ -1434,11 +1427,14 @@ def _set_slicing_by_batch_id(
         """
 
         for pid, policy_batch in batch.policy_batches.items():
-            if self.module[pid].is_stateful():
-                # We assume that arriving batches for recurrent modules are already
-                # padded to the max sequence length and have tensors of shape
-                # [B, T, ...]. Therefore, we slice sequence lengths in B. See
-                # SampleBatch for more information.
+            # We assume that arriving batches for recurrent modules OR batches that
+            # have a SEQ_LENS column are already zero-padded to the max sequence length
+            # and have tensors of shape [B, T, ...]. Therefore, we slice sequence
+            # lengths in B. See SampleBatch for more information.
+            if (
+                self.module[pid].is_stateful()
+                or policy_batch.get("seq_lens") is not None
```

Comment:
Bug fix: For APPO, we do have SEQ_LENS and need to slice on the B-axis (not T-axis), but the Module might still not be stateful.

Reply:
does `policy_batch.get("seq_lens") is not None` cover `self.module[pid].is_stateful()` or no?

Reply:
actually: not sure :)
I do think these are different concepts. You might have a stateful model that does NOT operate on (time series) sequences (e.g. DreamerV3; well, it does operate on sequences but does NOT need the `seq_lens` key) or vice-versa (like transformers, operating on sequences but statelessly).